tun2socks/intra/tcp.go

```diff
@@ -0,0 +1,78 @@
+// Copyright 2019 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Derived from go-tun2socks's "direct" handler under the Apache 2.0 license.
+
+package intra
+
+import (
+	"io"
+	"net"
+
+	"github.com/eycorsican/go-tun2socks/common/log"
+	"github.com/eycorsican/go-tun2socks/core"
+)
+
+type tcpHandler struct {
+	fakedns net.Addr
+	truedns net.Addr
+}
+
+// NewTCPHandler returns a TCP forwarder with Intra-style behavior.
+// Currently this class only redirects DNS traffic to a
+// specified server.  (This should be rare for TCP.)
+// All other traffic is forwarded unmodified.
+func NewTCPHandler(fakedns, truedns net.Addr) core.TCPConnHandler {
+	return &tcpHandler{fakedns: fakedns, truedns: truedns}
+}
+
+func (h *tcpHandler) handleUpload(local net.Conn, remote *net.TCPConn) {
+	// TODO: Handle half-closed sockets more correctly.
```

Comment:
Don't use Close(). Use CloseWrite() instead. See what I did in outline-ss-server:
https://github.com/Jigsaw-Code/outline-ss-server/blob/8972cd64904d445b27bfb7efd083f1d3bc5bbf29/net/net.go#L56

Reply:
Also, you can use a single `copyOneWay` to avoid duplication.
It would be nice to reuse the same pattern across our code for consistent behavior and to avoid errors.

Reply:
`local` is a `net.Conn`, which doesn't have `CloseWrite`.  I've clarified the TODO to explain why we can't do this yet.

Reply:
We discussed offline. The CloseWrite and CloseRead are helpful. The local connection we would need to cast to TCPConn, but I think it's a custom net.Conn that go-tun2socks defines, so we should look into fixing it there instead.